### ISSUE TITLE
[Proposal] Provide alternative service resolution mechanism besides mdns

### ISSUE THREAD

**Issue by wcs1only** [dapr/dapr#3256]
## In what area(s)?

/area runtime

## Describe the feature

We commonly get reports that service invocation does not work with dapr standalone mode:
https://github.com/dapr/dapr/issues/2665
https://github.com/dapr/quickstarts/issues/361
https://github.com/dapr/quickstarts/issues/388
https://github.com/dapr/quickstarts/issues/399 (May also be a general failure to start daprd?)
https://github.com/dapr/python-sdk/issues/146
... and a bunch more across various SDKs/docs repos

We've root caused most of these to security/VPN configurations blocking mdns requests. The fix varies heavily based on which OS/security software the user is using. It would be lovely to have a generic solution to this, either by falling back to another service discovery mechanism if mdns doesn't work, or replacing mdns outright with something else.

## Release Note

RELEASE NOTE:

**Comment by yaron2** [comment]
We can point users to use Consul in self hosted mode.

Alternatively, if we find another cross machine self discovery mechanism that doesnt involve an external agent (as this is already provided with Consul) we can consider that.

**Comment by wcs1only** [comment]
> We can point users to use Consul in self hosted mode.
> 
> Alternatively, if we find another cross machine self discovery mechanism that doesnt involve an external agent (as this is already provided with Consul) we can consider that.

I'm hoping for a solution a little more out of the box than "point users to this." This seems to be tripping up users on the quickstart hello-world path, and the last thing I'd want to do is introduce another thing to install/understand before getting started.

**Comment by markheath** [comment]
It would be really nice for there to be a simple alternative for self-hosted systems where this is an issue. There is something with our corporate network that is blocking service to service invocation. Every time I think I've worked out how to work around it, it comes back. I can imagine this could be a very frustrating issue for people trying out Dapr for the first time.

**Comment by artursouza** [comment]
@mukundansundar Any idea on this?

**Comment by pkedy** [comment]
Here are a few thoughts from an offline conversation:

Yes, Consul would absolutely work, but is another container that the CLI would have to install.

An alternative is to add "Consul lite" functionality to the placement service. This could range anywhere from a "local only mode" which holds a simple in-memory lookup table to a fully distributed/replicated lookup table via Raft. The placement service already handles replicating the actors placement table. In theory, adding general service addresses could make use of existing infra/replication code. One could ask "are you just recreating Consul"? Yeah sort of. That aside, assuming the goals are 1) lightweight/no additional installation footprint and 2) satisfies Dapr service invocation without the pitfalls of mDNS, then this approach could make sense.

/cc @artursouza

**Comment by wcs1only** [comment]
We might be overthinking this, we could just have a local resolver component that is just a flat file in ~/.dapr/

**Comment by wcs1only** [comment]
Alternatively, we could do whatever `dapr list` is doing.

**Comment by pkedy** [comment]
`dapr list` and the VS Code plugin both look at the running processes to discover app-id and ports. If we don't really need a distributed option, this makes a ton of sense and the code/logic already exists in the CLI.

**Comment by yaron2** [comment]
I don't support expanding the placement service to provide service discovery, although I am guilty of suggesting the same in the past. Assigning extra responsibilities for a component that is targeted at placing actors just because it's the nearest neighbor isn't something I can advocate for.

I do support creating a local name resolution component that's explicitly clear about providing dev/single-box discovery.

Using it would be as simple as configuring the following:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: appconfig
spec:
  nameResolution:
    component: "local"
```

The CLI could be configured to register the `local` component by default.

**Comment by mukundansundar** [comment]
For the local component, we could replicate similar logic to Consul but have it in a datastore for example Redis. Since we are already using Redis as a state store an an actor store we can use it for service resolution also I think. This way a service can self register on Init and perform GET requests on ResolveID calls. 

The potential issue in this is in the slim mode the fallback will still be mDNS or Redis needs to be stood up separately and used for service discovery.

One more concern is deregistration of the service once it is shutdown . Currently we do not have a onShutdown handle regarding name resolution for deregistration ... 

One option is extend the name resolution interface to handle onShutdown ... This will help with the current Consul component also.
The other option is to make use of Redis TTL for the keys and have the service periodically refresh the associated registration if possIble.

**Comment by pkedy** [comment]
@mukundansundar @yaron2 any concerns with using the `dapr list` approach? It would work in slim mode as well.

**Comment by yaron2** [comment]
I do not support using Redis or any state store for service discovery purposes as that would create a tight coupling between a state store and a name resolution component.

@pkedy I'm fine with a file based approach or whatever `dapr list` does.

**Comment by yaron2** [comment]
@artursouza that's my preferred approach then.

**Comment by aktxyz** [comment]
Can you all please make sure that this works with applications launched thru daprd directly (ie not just thru the dapr process).
Currently, `dapr list` does not list things started via daprd directly.

**Comment by artursouza** [comment]
> Can you all please make sure that this works with applications launched thru daprd directly (ie not just thru the dapr process). Currently, `dapr list` does not list things started via daprd directly.

There was a PR to fix that.

**Comment by aktxyz** [comment]
great ! thanks

**Comment by jigargandhi** [comment]
So I am planning to start work on this issue.

As per the discussions, I will create a new name resolution component `local` that imitates `dapr list` and does process name and argument parsing. This will make sure that if mdns is disabled on local machine then `local` name resolution component will work.

**Comment by aktxyz** [comment]
+1 !!!

**Comment by tony-clarke-amdocs** [comment]
I am facing the same issue. I had to log off VPN before I could get the first example to work. It took several hours even before I came to that point. Even now it's not clear to me how the VPN is causing issues. Is there a way to make mDNS work with the VPN on?

**Comment by badgeratu** [comment]
> > Can you all please make sure that this works with applications launched thru daprd directly (ie not just thru the dapr process). Currently, `dapr list` does not list things started via daprd directly.
> 
> There was a PR to fix that.

@artursouza Can you please provide a link to that PR? Because I don't believe `dapr list` is yet showing `daprd` instances started standalone:
![image](https://user-images.githubusercontent.com/5095203/146778314-d951421d-cec1-4635-ad8e-7b6c135bed2f.png)

**Comment by jjcollinge** [comment]
@wcs1only would using Gossip directly not be a viable alternative to mDNS? I think based on [this article](https://www.consul.io/docs/architecture/gossip) even Consul uses Gossip under the hood.

**Comment by safari137** [comment]
This may already be covered by another issue, but a work around for me works as follows:

1. Disconnect from VPN (Cisco AnyConnect in my case)
2. Start the apps with Dapr
3. Re-connect to VPN

This is pretty painful in development as I do restart the apps often as I make changes.  If my application didn't have to access the corporate network, that wouldn't be a big deal though.

I'm going to work on running dapr on WSL/ubuntu since I could at least script the startup/shutdown of the apps and vpn.

**Comment by addjuarez** [comment]
https://github.com/dapr/components-contrib/pull/1380

**Comment by MrCliff** [comment]
Configuring the name resolver to resolve only dapr sidecars on local computer would be much needed feature for us. We have quite opposite issues with dapr in development but this should fix them too. When developing in office environment where mDNS is allowed, Dapr resolves to services on other developers' computers. This is because mDNS packets are thrown to the same network which other computers can see.

We have tried blocking inbound mDNS packets on Windows Firewall but somehow they don't get blocked at all. If blocking outbound mDNS, the name resolution stops working completely.

**Comment by MrCliff** [comment]
We found a solution for our issues which is to use Consul. However, the [guide on Dapr docs](https://docs.dapr.io/reference/components-reference/supported-name-resolution/setup-nr-consul/) doesn't give simple and complete example how to use Consul as name resolver in local environment. Thankfully I found [this great article](https://markheath.net/post/dapr-consul-name-resolution) that contains simple enough instructions for that. You could include such an example on the Dapr Docs Consul page.

**Comment by xckai** [comment]
> We found a solution for our issues which is to use Consul. However, the [guide on Dapr docs](https://docs.dapr.io/reference/components-reference/supported-name-resolution/setup-nr-consul/) doesn't give simple and complete example how to use Consul as name resolver in local environment. Thankfully I found [this great article](https://markheath.net/post/dapr-consul-name-resolution) that contains simple enough instructions for that. You could include such an example on the Dapr Docs Consul page.

Thank you. It's been two years since this issue was raised, but Dapr has not yet provided an official solution.

**Comment by ItalyPaleAle** [comment]
Let's get this done for 1.13, for local development scenarios using SQLite.

**Comment by ItalyPaleAle** [comment]
New name resolver (backed by SQLite) is implemented in #7038

Whether that becomes the new default for the local DX is an issue that should be discussed with the runtime and CLI maintainers.

**Comment by msfussell** [comment]
@ItalyPaleAle Is this "Whether that becomes the new default for the local DX is an issue that should be discussed with the runtime and CLI maintainers." tracked as an issue?
